doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,192 @@
+
+.. flag:: Allow Rewrite Rules
+
+   Enables the use of rewrite rules.
+   It is disabled by default and cannot be disabled once enabled.
+   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+They stand in-between defined constants and axioms,
+in that they don't always reduce as defined constants do,
+but they may still reduce using the provided rules, unlike axioms.
+
+.. cmd:: {| Symbol | Symbols } {| @assumpt | {+ ( @assumpt ) } }
+   :name: Symbol; Symbols
+
+   Binds an :n:`@ident` to a :n:`@type` as a symbol.
+
+   .. coqtop:: in
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Rewrite rules
+---------------
+.. cmd:: Rewrite {| Rule | Rules } @ident := {? %| } {+| @rewrite_rule }
+  :name: Rewrite Rule; Rewrite Rules
+
+  .. insertprodn rewrite_rule rewrite_rule
+
+  .. prodn::
+     rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term
+
+Declares a named block of rewrite rules.
+
+Rewrite rules have two parts named pattern (left-hand side) and replacement (right-hand side).
+Patterns are a subclass of :n:`@term`\s described :ref:`below<Pattern syntax>`,
+while replacements are regular :n:`@term`\s,
+which can also refer to the pattern variables matched by the pattern with the :n:`?@name` syntax.
+When a rule is applied, the term is matched against the pattern,
+subterms aligned with pattern variables are collected
+and then substituted into the replacement, which is returned.
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+     | ?n ++ 0 ==> ?n
+     | ?n ++ S ?m ==> S (?n ++ ?m)
+     | 0 ++ ?n ==> ?n
+     | S ?n ++ ?m ==> S (?n ++ ?m).
+
+.. _Pattern syntax:
+
+Pattern syntax
+--------------
+
+Patterns are a subclass of :n:`@term`\s which are rigid enough to be matched against.
+Informally, they are terms with pattern variables (:n:`?@name`),
+where those may not appear on the left of applications or as the discriminee of a match or a primitive projection;
+furthermore a pattern may not have let-bindings, (co-)fixpoints or non-symbol constants.
+
+As a formal grammar, it is easier to understand them with the separation between head-pattern (:n:`@rw_head_pattern`)
+and eliminations (non-base-case constructions for :n:`@rw_pattern`):
+
+  .. prodn::
+     rw_pattern ::= @rw_head_pattern
+     | @rw_pattern {+ @rw_pattern_arg }
+     | @rw_pattern .( @qualid {? @univ_annot } )
+     | match @rw_pattern {? as @name } {? in @pattern } {? return @rw_pattern_arg } with {*| @pattern => @rw_pattern_arg } end
+     rw_head_pattern ::= @ident
+     | @qualid {? @univ_annot }
+     | fun ({+ @name } {? : @rw_pattern_arg}) => @rw_pattern_arg
+     | forall ({+ @name } {? : @rw_pattern_arg}), @rw_pattern_arg
```

Comment:
Based on the underlying grammar (`term_forall_or_fun`), I expect the parentheses are not required.  And maybe multiple binders will work.  OTOH if your grammar misses a few cases, I think we can live with that in something marked "experimental".  Precisely describing a subset of an existing grammar by hand is hard to get right.

![image](https://github.com/coq/coq/assets/1253341/2286224a-df4e-4ff3-b4b6-a2a8b7ee93d2)

![image](https://github.com/coq/coq/assets/1253341/a3b1cfdb-6c3d-4fb5-8157-47fbd7ab6732)

Reply:
I added the + for multiple binders, but I don't think it would be useful to be complete : patterns are parsed from the parsed representation of terms, so all constructions that get resolved to the basic ones during the initial parsing are allowed (destructuring let, notations, etc.). It would help no-one for all of this to be included I believe

Reply:
The other way to approach it is to just refer to the underlying nonterminals and then describe in text what parts of them are not supported for rewrite rules.  Hopefully you get a clear error message when you use anything that's not supported.  But I'll leave it up to your choice.